### DISCUSSION TITLE
Speed up hex meshing (multithreading)

### DISCUSSION THREAD

**Discussion by zero-infinit** [neperfepx/neper#103] [Q&A] [answered]
Working on a workstation with 96 cpu cores. When doing meshing, neper is working on 64 threads but using only one of the cores. Is there any way to speed up meshing by using all the cores parallely?

```
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.1.1
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt
Info   : Running on 64 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2020, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/zeqing/.neperrc').
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] n741976-id1.tess -elt hex -cl 0.0001 -format msh,vtk
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading tessellation...
Info   :     [i] Parsing file `n741976-id1.tess'...
Info   :     [i] Parsed file `n741976-id1.tess'.
Info   : Meshing...
Info   :   - Preparing... (cl = 0.0001) 100%

Info   :   - 3D meshing...   9%
```

![image](https://user-images.githubusercontent.com/78400787/106557738-6cffcf00-655d-11eb-90f6-98bf31f18f88.png)

**Comment by zero-infinit** [comment]
Thinking about using partition `-part` to minimize computation time. But the need libScotch library, which was not enabled when installing Neper. In this case, do I need to remove Neper, including the dependencies, and re-install?

**Reply by rquey** [reply]
`-part` will create a partition of the resulting mesh, which is not what you want. If you wanted to include a new dependency, you would have to re-configure (e.g. using `ccmake`), then recompile (`make`) and install again (`make install`). This would overwrite the previous version (assuming you ran `make install` the first time).

**Reply by zero-infinit** [reply]
Thanks for your reply! Will there be any way to combine the partitioned mesh together?

**Reply by rquey** [reply]
I'm not sure what you mean. What partitioning does is to assign partition numbers to the elements and nodes. It does not change the mesh connectivity. The aim is to be able to assign each partition to a computation unit, in the parallel (finite element) simulation. Partitioning is useful when the mesh is used in [FEPX](https://fepx.info).

**Comment by rquey** [comment, marked as answer]
Hex meshing was not multithreaded.

I have multithreaded the existing algorithm in the latest version (4.1.3-1), which decreases the CPU time by a factor of about 2 to 3, depending on the tessellation, meshing parameters and number of threads. It is still much slower than it could be with a more efficient algorithm...
